hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/LazyFileIterable.java

```diff
@@ -0,0 +1,116 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.uber.hoodie.common.util.collection;
+
+import com.uber.hoodie.common.util.SpillableMapUtils;
+import com.uber.hoodie.exception.HoodieException;
+import com.uber.hoodie.exception.HoodieIOException;
+import org.apache.avro.Schema;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Iterable to lazily fetch values spilled to disk.
+ * This class uses RandomAccessFile to randomly access the position of
+ * the latest value for a key spilled to disk and returns the result.
+ * @param <T>
+ */
+public class LazyFileIterable<T> implements Iterable<T> {
+
+  // Used to access the value written at a specific position in the file
+  private RandomAccessFile readOnlyFileHandle;
+  // Stores the key and corresponding value's latest metadata spilled to disk
+  private Map<T, DiskBasedMap.ValueMetadata> inMemoryMetadataOfSpilledData;
+  // Schema used to de-serialize payload written to disk
+  private Schema schema;
+  // Class used to de-serialize/realize payload written to disk
+  private String payloadClazz;
+
+  public LazyFileIterable(RandomAccessFile file, Map<T, DiskBasedMap.ValueMetadata> map,
+                          Schema schema, String payloadClazz) {
+    this.readOnlyFileHandle = file;
+    this.inMemoryMetadataOfSpilledData = map;
+    this.schema = schema;
+    this.payloadClazz = payloadClazz;
+  }
+  @Override
+  public Iterator<T> iterator() {
+    try {
+      return new LazyFileIterator<>(readOnlyFileHandle, inMemoryMetadataOfSpilledData, schema, payloadClazz);
+    } catch(IOException io) {
+      throw new HoodieException("unable to initialize iterator for file on disk", io);
+    }
+  }
+
+  /**
+   * Iterator implementation for the iterable defined above.
+   * @param <T>
+   */
+  public class LazyFileIterator<T> implements Iterator<T> {
+
+    private RandomAccessFile readOnlyFileHandle;
+    private Schema schema;
+    private String payloadClazz;
+    private Iterator<Map.Entry<T, DiskBasedMap.ValueMetadata>> metadataIterator;
+
+    public LazyFileIterator(RandomAccessFile file, Map<T, DiskBasedMap.ValueMetadata> map,
+                            Schema schema, String payloadClazz) throws IOException {
+      this.readOnlyFileHandle = file;
+      this.schema = schema;
+      this.payloadClazz = payloadClazz;
+      // sort the map in increasing order of offset of value so disk seek is only in one(forward) direction
+      this.metadataIterator = map
```

Comment:
@vinothchandar here

Reply:
how much this improve the runtime by?  compared to cost of sorting?

Reply:
Here are the numbers from sorting vs non-sorting : 

----------------- With Sorting enabled---------------------------
18/02/20 20:21:05 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records 43588
18/02/20 20:21:05 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in memory 17109
18/02/20 20:21:05 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in disk 26479
18/02/20 20:21:05 INFO collection.LazyFileIterable: ExtraLogs:: Time Taken to sort 18
18/02/20 20:21:14 INFO compact.HoodieRealtimeTableCompactor: ExtraLogs::Time take to read from spillabel 8513

18/02/21 04:52:40 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records 187068
18/02/21 04:52:40 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in memory 16341
18/02/21 04:52:40 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in disk 170727
18/02/21 04:52:40 INFO collection.LazyFileIterable: ExtraLogs:: Time Taken to sort 146
18/02/21 04:53:45 INFO compact.HoodieRealtimeTableCompactor: ExtraLogs::Time take to read from spillabel 64863


----------------- Without Sorting ---------------------------
18/02/20 22:47:11 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records 170444
18/02/20 22:47:11 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in memory 16341
18/02/20 22:47:11 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in disk 154103
18/02/20 22:47:11 INFO collection.LazyFileIterable: ExtraLogs:: Time Taken to sort 0
18/02/20 22:48:26 INFO compact.HoodieRealtimeTableCompactor: ExtraLogs::Time take to read from spillabel 74631

18/02/21 19:04:38 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records 242087
18/02/21 19:04:38 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in memory 17740
18/02/21 19:04:38 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Size in memory 2900694448
18/02/21 19:04:38 INFO log.HoodieCompactedLogRecordScanner: ExtraLogs:: Number of records in disk 224347
18/02/21 19:04:38 INFO collection.LazyFileIterable: ExtraLogs:: Time Taken to sort 0
18/02/21 19:06:01 INFO compact.HoodieRealtimeTableCompactor: ExtraLogs::Time take to read from spillabel 82955

It seems like sorting gives a minor benefit but again this is controlled by the layout of the dataset on DISK in both cases and how many back and forth disk seeks the non-sorting actually does.
But I realized that this will not matter at the moment since we started to pass Map<> to HoodieMergeHandle instead of Iterator. So when we think of SortMerge we can spend more time on this, WDYT ?

Reply:
sure. gains also depend on how IO loaded the box really is or if the seek was teh bottleneck.. Even without sorting, its only taking < 1 ms to read, so def not seeking everytime.. and seems like page cache is kicking in.. 
This will come in handy on a loaded system anyways.. lets revisit down the line .. sure